Dalamud/Interface/Internal/TextureManager.cs

```diff
@@ -0,0 +1,458 @@
+
+namespace Dalamud.Interface.Internal;
+
+/// <summary>
+/// Service responsible for loading and disposing ImGui texture wraps.
+/// </summary>
+[PluginInterface]
+[InterfaceVersion("1.0")]
+[ServiceManager.BlockingEarlyLoadedService]
+#pragma warning disable SA1015
+[ResolveVia<ITextureSubstitutionProvider>]
+#pragma warning restore SA1015
+internal class TextureManager : IDisposable, IServiceType, ITextureSubstitutionProvider
+{
+    private const string IconFileFormat = "ui/icon/{0:D3}000/{1}{2:D6}.tex";
+    private const string HighResolutionIconFileFormat = "ui/icon/{0:D3}000/{1}{2:D6}_hr1.tex";
+
+    private const uint MillisecondsEvictionTime = 2000;
+    
+    private static readonly ModuleLog Log = new("TEXM");
+
+    private readonly Framework framework;
+    private readonly DataManager dataManager;
+    private readonly DalamudStartInfo startInfo;
+    
+    private readonly Dictionary<string, TextureInfo> activeTextures = new();
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="TextureManager"/> class.
+    /// </summary>
+    /// <param name="framework">Framework instance.</param>
+    /// <param name="dataManager">DataManager instance.</param>
+    /// <param name="startInfo">DalamudStartInfo instance.</param>
+    [ServiceManager.ServiceConstructor]
+    public TextureManager(Framework framework, DataManager dataManager, DalamudStartInfo startInfo)
+    {
+        this.framework = framework;
+        this.dataManager = dataManager;
+        this.startInfo = startInfo;
+
+        this.framework.Update += this.FrameworkOnUpdate;
+    }
+
+    /// <inheritdoc/>
+    public event ITextureSubstitutionProvider.TextureDataInterceptorDelegate? InterceptTexDataLoad;
+    
+    /// <summary>
+    /// Get a texture handle for a specific icon.
+    /// </summary>
+    /// <param name="iconId">The ID of the icon to load.</param>
+    /// <param name="flags">Options to be considered when loading the icon.</param>
+    /// <param name="language">
+    /// The language to be considered when loading the icon, if the icon has versions for multiple languages.
+    /// If null, default to the game's current language.
+    /// </param>
+    /// <param name="keepAlive">
+    /// Prevent Dalamud from automatically unloading this icon to save memory. Usually does not need to be set.
+    /// </param>
+    /// <returns>
+    /// Null, if the icon does not exist in the specified configuration, or a texture wrap that can be used
+    /// to render the icon.
+    /// </returns>
+    public TextureManagerTextureWrap? GetIcon(uint iconId, ITextureProvider.IconFlags flags = ITextureProvider.IconFlags.ItemHighQuality, ClientLanguage? language = null, bool keepAlive = false)
+    {
+        var hiRes = flags.HasFlag(ITextureProvider.IconFlags.HiRes);
+        
+        // 1. Item
+        var path = FormatIconPath(
+            iconId,
+            flags.HasFlag(ITextureProvider.IconFlags.ItemHighQuality) ? "hq/" : string.Empty,
```

Comment:
At the moment, items will not fall back to non-hq if there is no hq variant and the hq flag is set. Do we want that?

Reply:
I think no fallback for HQ request is fine. We _may_ consider a fallback for HR, but that may be pretty niche and not worth doing.